onnxruntime/core/session/inference_session.cc

```diff
@@ -873,6 +878,20 @@ common::Status InferenceSession::Initialize() {
     // re-acquire mutex
     std::lock_guard<onnxruntime::OrtMutex> l(session_mutex_);
 
+    // At this time we know all the providers that will be part of this session.
+    // Read shared allocators from the environment and update them in the respective providers.
+    //
+    // The reason for updating the providers is so that when the session state is created the allocators
+    // are setup appropariately keyed by OrtMemoryInfo with delegates going to the respective providers.
+    // Secondly, the GetAllocator() method inside IExecutionProvider is still used in various places, hence
+    // it doesn't make sense to just update the allocator map inside session state with these shared allocators; doing
+    // so would cause inconsistency between the allocator map inside session sate and that inside the providers.
+    // TODO: we could refactor the allocators to not require the call to GetAllocator but that change is much bigger
+    // since we've to take into account the per-thread cuda allocators.
+    if (session_options_.use_shared_allocator) {
```

Comment:
Is it worth noting here that we can possibly make the change cleaner if the per-thread logic is layered on top as in IAllocator (similar to how the arena is layered on top of the device allocator)?